File: app/src/main/java/com/paxus/pay/poslinkui/demo/view/SelectOptionsView.java
package com.paxus.pay.poslinkui.demo.view;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.Context;
import android.graphics.Typeface;
import android.text.Spannable;
import android.text.SpannableStringBuilder;
import android.text.style.AbsoluteSizeSpan;
import android.text.style.StyleSpan;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.InputMethodManager;
import android.widget.CheckedTextView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.paxus.pay.poslinkui.demo.R;
import com.paxus.pay.poslinkui.demo.utils.Logger;

import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class SelectOptionsView extends RecyclerView {
    private boolean isInitialized = false, isModified = false;
    private OptionAdapter adapter;
    private Set<SelectOptionsView> siblings = new HashSet<>();
    private Activity context;

    public SelectOptionsView(@NonNull Context context) {
        super(context);
    }

    public SelectOptionsView(@NonNull Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
    }

    public SelectOptionsView(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    public void initialize(Activity context, int columnCount, List<Option> options, OptionSelectListener listener) {
        this.adapter = new OptionAdapter(context, options, listener);
        setLayoutManager(new GridLayoutManager(context, columnCount));
        this.context = context;
        this.isInitialized = true;
        this.setAdapter(this.adapter);
    }

    public void clear(){
        if(!isInitialized){
            Logger.e(getClass().getSimpleName() + " not yet initialized. " + new Object(){}.getClass().getEnclosingMethod().getName() + " cannot be executed.");
            return;
        }

        if(isModified){
            this.adapter.unselectAllOptions((ViewGroup) this.getRootView());
            this.isModified = false;
        }

        if(!siblings.isEmpty()){
            for(SelectOptionsView sibling : siblings) {
                if(sibling.isModified) sibling.clear();
            }
        }
    }

    public void append(SelectOptionsView selectOptionsView) {
        if(!selectOptionsView.isInitialized) {
            Logger.e(selectOptionsView.getClass().getSimpleName() + " not yet initialized. " + new Object(){}.getClass().getEnclosingMethod().getName() + " cannot be executed.");
            return;
        }

        for(SelectOptionsView sibling : this.siblings) {
            sibling.addSibling(selectOptionsView);
            selectOptionsView.addSibling(sibling);
        }
        this.siblings.add(selectOptionsView);
        selectOptionsView.addSibling(this);
    }

    private void addSibling(SelectOptionsView selectOptionsView) {
        this.siblings.add(selectOptionsView);
    }

    public static class Option {
        Integer index;
        String title, subtitle;
        Object value;

        public Option(Integer index, String title, String subtitle, Object value) {
            this.index = index;
            this.title = title;
            this.subtitle = subtitle;
            this.value = value;
        }

        public Object getValue() {
            return value;
        }
    }

    @FunctionalInterface
    public interface OptionSelectListener {
        void onSelect(Option option);
    }

    private class OptionAdapter extends RecyclerView.Adapter<OptionAdapter.OptionHolder> {
        private List<Option> options;
        private LayoutInflater layoutInflater;
        private OptionSelectListener optionSelectListener;
        private boolean isDoubleHeightRequired = false;

        OptionAdapter(Context context, List<Option> options, OptionSelectListener optionSelectListener) {
            this.options = options;
            this.layoutInflater = LayoutInflater.from(context);
            this.optionSelectListener = optionSelectListener;
            for(Option option : this.options) {
                if(option.title != null && option.subtitle != null) {
                    isDoubleHeightRequired = true;
                    break;
                }
            }
        }

        @NonNull
        @Override
        public OptionHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
            View holder = layoutInflater.inflate(R.layout.layout_select_option_item, parent, false);
            if(isDoubleHeightRequired){
                ViewGroup.LayoutParams layoutParams = holder.getLayoutParams();
                layoutParams.height = (int) getResources().getDimension(R.dimen.button_height_double_line);
                holder.setLayoutParams(layoutParams);
            }
            return new OptionHolder(holder);
        }

        @Override @SuppressLint("SetTextI18n")
        public void onBindViewHolder(@NonNull OptionHolder holder, int position) {
            Option option = options.get(position);
            SpannableStringBuilder text = new SpannableStringBuilder();

            if (option.title != null) {
                text.append(option.title);
                text.setSpan(new AbsoluteSizeSpan((int) getResources().getDimension(R.dimen.text_size_title)), 0, option.title.length(), Spannable.SPAN_INCLUSIVE_INCLUSIVE);
            }
            if (option.subtitle != null) {
                text.append(option.title != null ? "\n" : "").append(option.subtitle);
                text.setSpan(new AbsoluteSizeSpan((int) (option.title != null ? getResources().getDimension(R.dimen.text_size_hint) : getResources().getDimension(R.dimen.text_size_title))),
                        text.length() - option.subtitle.length(), text.length(), Spannable.SPAN_INCLUSIVE_INCLUSIVE);
                text.setSpan(new StyleSpan(Typeface.BOLD), 0, option.title.length(), Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);
            }

            holder.text.setText(text);
            holder.bindListener(option, optionSelectListener);
        }

        @Override
        public int getItemCount() {
            return options.size();
        }

        private void unselectAllOptions(ViewGroup parent) {
            for (int i = 0; i < parent.getChildCount(); i++) {
                if (parent.getChildAt(i) instanceof CheckedTextView)
                    ((CheckedTextView) parent.getChildAt(i)).setChecked(false);
                else if (parent.getChildAt(i) instanceof ViewGroup)
                    unselectAllOptions((ViewGroup) parent.getChildAt(i));
            }
        }

        private class OptionHolder extends RecyclerView.ViewHolder {
            public TextView text;

            public OptionHolder(@NonNull View itemView) {
                super(itemView);
                text = itemView.findViewById(R.id.layout_select_option_item_text);
                text.setTextAlignment(View.TEXT_ALIGNMENT_CENTER);
            }

            public void bindListener(final Option option, final OptionSelectListener listener) {
                itemView.setOnClickListener(view -> {
                    InputMethodManager inputMethodManager = (InputMethodManager) getContext().getSystemService(Context.INPUT_METHOD_SERVICE);
                    inputMethodManager.hideSoftInputFromWindow(context.getWindow().getDecorView().getWindowToken(), InputMethodManager.HIDE_NOT_ALWAYS);
                    listener.onSelect(option);
                    clear();
                    ((CheckedTextView) text).setChecked(true);
                    isModified = true;
                });
            }
        }

    }
}

File: app/src/main/java/com/paxus/pay/poslinkui/demo/view/SelectOptionsView.kt
package com.paxus.pay.poslinkui.demo.view

import android.annotation.SuppressLint
import android.app.Activity
import android.content.Context
import android.graphics.Typeface
import android.text.Spannable
import android.text.SpannableStringBuilder
import android.text.style.AbsoluteSizeSpan
import android.text.style.StyleSpan
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.CheckedTextView
import android.widget.TextView
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.paxus.pay.poslinkui.demo.R
import com.paxus.pay.poslinkui.demo.utils.Logger
import com.paxus.pay.poslinkui.demo.view.SelectOptionsView.OptionAdapter.OptionHolder

class SelectOptionsView : RecyclerView {
    private var isInitialized = false
    private var isModified = false
    private var optionAdapter: OptionAdapter? = null
    private val siblings: MutableSet<SelectOptionsView> = HashSet()

    constructor(context: Context) : super(context)

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs)

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(
        context,
        attrs,
        defStyleAttr
    )

    fun initialize(
        activity: Activity?,
        columnCount: Int,
        options: MutableList<Option>,
        listener: OptionSelectListener
    ) {
        optionAdapter = OptionAdapter(activity, options, listener)
        layoutManager = GridLayoutManager(activity, columnCount)
        isInitialized = true
        super.setAdapter(optionAdapter)
    }

    fun clear() {
        if (!isInitialized) {
            Logger.e(
                "${javaClass.simpleName} not yet initialized; clear() cannot be executed."
            )
            return
        }

        if (isModified) {
            val ad = optionAdapter
            val root = this.rootView as? ViewGroup
            if (ad != null && root != null) ad.unselectAllOptions(root)
            this.isModified = false
        }

        if (siblings.isNotEmpty()) {
            for (sibling in siblings) {
                if (sibling.isModified) sibling.clear()
            }
        }
    }

    fun append(selectOptionsView: SelectOptionsView) {
        if (!selectOptionsView.isInitialized) {
            Logger.e(
                "${selectOptionsView.javaClass.simpleName} not yet initialized; append() cannot be executed."
            )
            return
        }

        for (sibling in this.siblings) {
            sibling.addSibling(selectOptionsView)
            selectOptionsView.addSibling(sibling)
        }
        this.siblings.add(selectOptionsView)
        selectOptionsView.addSibling(this)
    }

    private fun addSibling(selectOptionsView: SelectOptionsView?) {
        selectOptionsView?.let { this.siblings.add(it) }
    }

    class Option(var index: Int?, var title: String?, var subtitle: String?, var value: Any?)

    fun interface OptionSelectListener {
        fun onSelect(option: Option?)
    }

    private inner class OptionAdapter(
        inflationContext: Context?,
        private val options: MutableList<Option>,
        private val optionSelectListener: OptionSelectListener
    ) : Adapter<OptionHolder>() {
        private val layoutInflater: LayoutInflater
        private var isDoubleHeightRequired = false

        init {
            this.layoutInflater = LayoutInflater.from(inflationContext)
            for (option in this.options) {
                if (option.title != null && option.subtitle != null) {
                    isDoubleHeightRequired = true
                    break
                }
            }
        }

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): OptionHolder {
            val holder = layoutInflater.inflate(R.layout.layout_select_option_item, parent, false)
            if (isDoubleHeightRequired) {
                val layoutParams = holder.layoutParams
                layoutParams.height =
                    resources.getDimension(R.dimen.button_height_double_line).toInt()
                holder.layoutParams = layoutParams
            }
            return OptionHolder(holder)
        }

        @SuppressLint("SetTextI18n")
        override fun onBindViewHolder(holder: OptionHolder, position: Int) {
            val option = options[position]
            val text = SpannableStringBuilder()

            val titleText = option.title
            val subtitleText = option.subtitle
            if (titleText != null) {
                text.append(titleText)
                text.setSpan(
                    AbsoluteSizeSpan(
                        resources.getDimension(R.dimen.text_size_subtitle).toInt()
                    ), 0, titleText.length, Spannable.SPAN_INCLUSIVE_INCLUSIVE
                )
            }
            if (subtitleText != null) {
                text.append(if (titleText != null) "\n" else "").append(subtitleText)
                val subtitleSizePx = if (titleText != null) {
                    resources.getDimension(R.dimen.text_size_hint)
                } else {
                    resources.getDimension(R.dimen.text_size_subtitle)
                }
                text.setSpan(
                    AbsoluteSizeSpan(subtitleSizePx.toInt()),
                    text.length - subtitleText.length,
                    text.length,
                    Spannable.SPAN_INCLUSIVE_INCLUSIVE
                )
                if (titleText != null) {
                    text.setSpan(
                        StyleSpan(Typeface.BOLD),
                        0,
                        titleText.length,
                        Spannable.SPAN_EXCLUSIVE_EXCLUSIVE
                    )
                }
            }

            holder.text.text = text
            holder.bindListener(option, optionSelectListener)
        }

        override fun getItemCount(): Int {
            return options.size
        }

        fun unselectAllOptions(parent: ViewGroup) {
            for (i in 0 until parent.childCount) {
                val child = parent.getChildAt(i)
                when (child) {
                    is CheckedTextView -> child.isChecked = false
                    is ViewGroup -> unselectAllOptions(child)
                    else -> Unit
                }
            }
        }

        private inner class OptionHolder(itemView: View) : ViewHolder(itemView) {
            var text: TextView

            init {
                text = itemView.findViewById<TextView>(R.id.layout_select_option_item_text)
                text.textAlignment = TEXT_ALIGNMENT_CENTER
            }

            fun bindListener(option: Option?, listener: OptionSelectListener) {
                itemView.setOnClickListener {
                    listener.onSelect(option)
                    clear()
                    (text as CheckedTextView).isChecked = true
                    isModified = true
                }
            }
        }
    }

    companion object {
        fun buildOptions(options: Array<String?>?): MutableList<Option?> {
            val list: MutableList<Option?> = ArrayList<Option?>()
            if (options == null) return list

            for (i in options.indices) {
                list.add(Option(null, options[i], null, i))
            }
            return list
        }
    }
}
